Add framework to write more robust action scheduler based migrations.

vedanshujain opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

From time to time, we have had the need to migrate a lot of data, either to power features or to fix bugs. However, we don’t have a shared mechanism to do this, so every developer working on implementing any kind of migration would do it from scratch. Since the developer implements it from scratch, we often also push bugs specific to the migration routine itself that we have previously solved before.

In this PR, I propose implementing a lightweight framework to write a migration that we can use most of the time. This framework will consider meta aspects of migration such as scheduling, re-entring, pausing, resuming, etc. We, anyway, need to work on [making migrations more robust](https://github.com/woocommerce/woocommerce/issues/32922) as part of the Custom order tables project, so it’s a good time to make it re-usable as we implement it.

We can create two classes – `BatchProcessingController` and`BatchProcessor`, where:

1. `BatchProcessor` will be an abstract class that will provide a framework for implementing details specific to migration. Such as methods to process batches. All the meta details like starting the process, error handling, updating progress status, etc., will be already implemented.
1. `BatchProcessingController` class will take care of running the migration as needed, starting again if it gets stuck, scheduling in the action scheduler queue, etc. This also implements a periodic check for all pending migration and schedules action for error’ed migrations as well.

In other words, we divide actual migration logic and logic for scheduling migration into two separate concerns. While the actual migration logic changes on case to case basis, the logic for scheduling the migrations can remain the same.

This PR also modifies custom order table migrations to use this common framework.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #32922 .

### How to test the changes in this Pull Request:

The easiest way to test is to run the COT migration from the interface -

1. Go to WooCommerce > Status > Tools and delete and recreate custom order tables.
1. Go to WooCommerce > Settings > Advanced > Custom Data Stores and start the migration process.
1. Verify that migration starts and completes as expected.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

vedanshujain committed:
Add framework to write more robust action scheduler based migrations.

vedanshujain committed:
Add framework for robust Action based migrations.

vedanshujain committed:
DataSynchronizer for COT to use base class for WCActionUpdator.

botwoo commented:
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33233/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Add framework for robust Action based migrations.` c2c23726ac7b9384c3df495b537cc36fcc87804b
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

vedanshujain committed:
Refactor class names from WCUpdates to Batch

vedanshujain committed:
Update DataSynchronize with latest refactors in batchprocessor class.

vedanshujain committed:
Remove ID function in favor of classname.

vedanshujain committed:
Add tests for batchprocessingcontroller, also share concrete to prevent duplicating actions.

vedanshujain committed:
Pass classname directly since ->id method was removed.

vedanshujain committed:
Add changelog.

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php:
```diff
@@ -0,0 +1,200 @@
+	const PENDING_PROCESS_OPTION_NAME = 'wc_pending_action_updates';
+	const CONTROLLER_CRON_NAME = 'wc_schedule_pending_updates';
+	const SINGLE_PROCESS_MIGRATION_ACTION = 'wc_run_action_update';
+	const ACTION_GROUP = 'wc_updates';
```
Since we are moving to the concept of "batches", I suggest renaming the options and actions as folloes:

```suggestion
	const PENDING_PROCESS_OPTION_NAME = 'wc_pending_batch_updates';
	const CONTROLLER_CRON_NAME = 'wc_schedule_pending_batch_updates';
	const SINGLE_PROCESS_BATCH_ACTION = 'wc_run_batch_update';
	const ACTION_GROUP = 'wc_batch_updates';
```

vedanshujain replied:
Thanks, added in https://github.com/woocommerce/woocommerce/pull/33233/commits/e922be012e20ff76dc1fc8357d357b573f8c4cd2#diff-8989ad7f50c6e63bf0ff83b5df17d72121166df76e16875ec985062b8ca740d5R17-R20

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php:
```diff
@@ -0,0 +1,200 @@
+	 */
+	public function __construct() {
+		add_action( self::CONTROLLER_CRON_NAME, array( $this, 'schedule_processes' ) );
+		add_action( self::SINGLE_PROCESS_MIGRATION_ACTION, array( $this, 'process_single_update' ), 10, 2 );
```
In line with what's done in other recent parts of the code, I suggest making the target methods private and replacing the `array($this,'method')` with a callback, `function() {$this->method();}`.

vedanshujain replied:
Hmm, not sure about this one. Wouldn't this cause problem with folks try to remove the hook for example?

Konamiman replied:
I've been using this strategy for all the new code I've created in `src` for quite a few time already, and so far we've only got a couple of instances of people complaining about not being able to remove the hook. In return we get cleaner code with a reduced public surface (only the methods that are truly intended to be `public` are), also an interesting question would be _why_ would anyone want to remove these hooks (as doing that would basically break the intended functionality).

Also a private method can be turned public in the future if we really need to do so, but the other way around is not possible.

See also: pcShBQ-hc-p2#comment-782

vedanshujain replied:
Ah, I see. Addressed in https://github.com/woocommerce/woocommerce/pull/33233/commits/143c5ed5a782849d0c83c3e871a3cec8366fb7a2#diff-77d9531091cce5f175d5bb8fe3a1560d9dda936d370c81fabb90ca0f46633e52R26-R39

Konamiman commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -227,6 +228,7 @@ private function init_hooks() {
 		wc_get_container()->get( RestockRefundedItemsAdjuster::class );
 		wc_get_container()->get( CustomOrdersTableController::class );
 		wc_get_container()->get( OptionSanitizer::class );
+		wc_get_container()->get( BatchProcessingController::class );
```
Not really related to this pull request, but I think an easy optimization we can do here is to cache the container in a variable once and then use it for all the `get`s.

vedanshujain replied:
Shouldn't this already be cached, this is just a proxy to `$GLOBALS['wc_container']`

Konamiman replied:
Sure, but to know that we have to look into how the function is implemented, and that implementation could change in the future. Additionally I think a "batch" of `$container->...` is more readable than one of `wc_get_container()->...`

vedanshujain replied:
Added in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-97fb927371ee663e98afaf8b8134905a96d4b59c68f6a83c569ec42a6b468a08R223-R232

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php:
```diff
@@ -0,0 +1,200 @@
+		 * @since 6.7.0.
+		 */
+		return apply_filters( 'woocommerce_get_batch_processor', $processor, $updater_class_name );
+	}
```
This is the way to go to get an instance of the processor, but I'd add two more steps:

1. If neither the container nor the filter can provied an instance of the processor, try to instantiate the class with `new` as a last resort.
2. If the above fails, explicitly throw an exception like "Can't get an instance of the batch processor X"; the code would fail shortly after anyway with a not so clear null ref exception.

vedanshujain replied:
👍  added in https://github.com/woocommerce/woocommerce/pull/33233/commits/e922be012e20ff76dc1fc8357d357b573f8c4cd2#diff-8989ad7f50c6e63bf0ff83b5df17d72121166df76e16875ec985062b8ca740d5R205-R212

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php:
```diff
@@ -0,0 +1,200 @@
+	 *
+	 * @return bool Whether the update is in progress.
+	 */
+	public function is_batch_process_pending( string $process_id ) : bool {
```
Name change suggestion:

```suggestion
	public function is_enqueued( string $processor_class_name ) : bool {
```

vedanshujain replied:
👍  Added in https://github.com/woocommerce/woocommerce/pull/33233/commits/b41f42203027d1654a722ce0b4af7e16f1d5cf94

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php:
```diff
@@ -0,0 +1,200 @@
+	/**
+	 * Stops all pending updates, and clears current memory state.
+	 */
+	public function force_clear_all_processes() {
```
Name change suggestion:

```suggestion
	public function cancel_queue() {
```

vedanshujain replied:
So this does a bit more than canceling queue, as it will also clear the cron process that manages everything and not just the processors. That's I wanted this method to have bit more aggressive name.

Konamiman replied:
Ok, then maybe I would have gone for something more succint as `cancel_all`, but I won't argue you if you are willing to get aggressive. 😄

vedanshujain replied:
😅, let's keep it as it is for now then.

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php:
```diff
@@ -0,0 +1,200 @@
+	 *
+	 * @return bool Whether it's already scheduled.
+	 */
+	public function already_scheduled( string $update_name ) {
```
Name change suggestion:

```suggestion
	public function is_enqueued( string $processor_class_name ) {
```

vedanshujain replied:
Note that there is already a method `is_enqueued` that checks if the process is registered and will be scheduled. This method is an abstraction for `as_has_scheduled_action` which indicates if a batch is going to be processed very soon.
Updated the param name in https://github.com/woocommerce/woocommerce/pull/33233/commits/e922be012e20ff76dc1fc8357d357b573f8c4cd2#diff-8989ad7f50c6e63bf0ff83b5df17d72121166df76e16875ec985062b8ca740d5R185

Konamiman replied:
So enqueued !== scheduled, ok, I get it. In that case I'd go with `is_scheduled` for consistency and readability.

vedanshujain replied:
👍 , renamed in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-77d9531091cce5f175d5bb8fe3a1560d9dda936d370c81fabb90ca0f46633e52R197

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php:
```diff
@@ -0,0 +1,200 @@
+	 *
+	 * @param string $updater_class_name Fully qualified class name of an updater, must be child class of `BatchProcessor`.
+	 */
+	public function enqueue_processor( string $updater_class_name ) {
```
Parameter name change suggestion:

```suggestion
	public function enqueue_processor( string $processor_class_name ) {
```

vedanshujain replied:
👍  Added in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-8989ad7f50c6e63bf0ff83b5df17d72121166df76e16875ec985062b8ca740d5R35

Konamiman commented on plugins/woocommerce/src/Internal/Updates/BatchProcessor.php:
```diff
@@ -0,0 +1,151 @@
+	 */
+	protected function get_process_option_name() {
+		// Option name max length is 191 chars.
+		return 'wc_batch_' . substr( self::class, -182 );
```
I'd rather throw an exception here, since using a truncated class name can potentially lead to duplicate names. But anyway, with the structural changes I have suggested this won't be needed anymore.

vedanshujain replied:
I have modified this so that it won't have conflict in https://github.com/woocommerce/woocommerce/pull/33233/commits/e922be012e20ff76dc1fc8357d357b573f8c4cd2#diff-8989ad7f50c6e63bf0ff83b5df17d72121166df76e16875ec985062b8ca740d5R143-R147. We still need it though to store the process state.

vedanshujain committed:
Refactor batch processor to make processor class lighter.

vedanshujain committed:
Update tests for modified logic.

vedanshujain committed:
Fine tuning scheduling parent cron and individual actions.

Parent cron manager should be scheduled only after there is atleast one update needed to be scheduled. First cron acitons schedules immediately, afterwards it's only every hour.

vedanshujain committed:
Rename method for better clarity.

vedanshujain committed:
Modified class names for clarity.

vedanshujain reviewed (commented):
> Moving responsibilities

Thanks for the suggestion, I have moved the methods around, BatchProcessor should be pretty lightweight now (and is stateless).

> Enqueueing vs registering

Enqueuing works like registering as you suggest. `BatchProcessingController` class is always aware of pending and scheduled processes. To maintain performance, the `always-on` action only starts when a batch process is enqueued for the first time, and when everything is done it also goes away.

Since it's maintaining a state (and saving it in an option) we can show all pending actions in a nice UI later on if we want to.

> Additional hooks

> wc_batch_processing_batch_size

I am not sure about this, because the way I am imagining this, we will control the batch size ourselves depending upon how much time a process eventually. Default batch size is just a suggestion to start with, adding a filter complicates it as we won't have control over that algorithm in the future.

> wc_batch_processing_finished

Let's add this one when we need it, right now we have a method that is called when a batch completes. I am hoping that we won't need this action because there is already a method which will be called in the BatchProcessor class when it's completed.

Konamiman commented on plugins/woocommerce/src/Internal/Utilities/BatchProcessor.php:
```diff
@@ -0,0 +1,66 @@
+	 *
+	 * @return array Batch of records.
+	 */
+	abstract public function get_batch_data( int $size, $last_processed ) : array;
```
About the method name: I still thing that `get_next_batch_to_process` better conveys the purpose of the method, as `get_batch_data` sounds too generic (batch data? Which batch data?)

About the `$last_processed` argument: I get the intent, you want to shield the processor class from having to store any kind of intermediate state. However I think there's a better way to achieve this: a `&$state` parameter passed by reference. Pass it as `null` the first time, let the processor set/alter it as it wishes, and just store it treating it as an opaque value. Bonus points if you pass the same parameter to `process_for_batch` and to `get_total_pending_count` too.

vedanshujain replied:
I have changed the method name in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-1e070cc170d3b65cd6ba61b60cbd0697c8431fda674b4720926c8fbc1afecce7R297. I agree about the state param, but then we would have to manage it's safe serialization and vice-versa, so let's add it when we need it.

Konamiman replied:
In that case maybe we can just remove the parameter, since:

1. We don't know what the processors will return for processing (each item of a "batch" can be anything) so we'll have the same problem with (de)serialization.
2. The processor itself can store any needed state data by itself (same reasoning as in the case of the cleanup after no more items are left to process).
3. The only processor instance we have so far, the one for orders, doesn't use it at all (and this is a good example of point 2).

So I think that either we provide an opaque state object that we pass around, dealing with any (de)serialization that might be needed, or we don't provide anything at all. That "last processed" argument feels like we are assuming how the processor behaves (what if the batch data it returns, or the processing it performs, isn't sequential?)

vedanshujain replied:
I agree and lean more towards not having a state to pass at all. I have removed the `last_processed` argument in https://github.com/woocommerce/woocommerce/pull/33233/commits/311f014dc39a5cfe7be85e3eef6c332cddae4618

Konamiman commented on plugins/woocommerce/src/Internal/Utilities/BatchProcessor.php:
```diff
@@ -0,0 +1,66 @@
+	 *
+	 * @param array $batch Batch details.
+	 */
+	abstract public function process_for_batch( array $batch );
```
This method name may be confusing: "process for batch" can be interpreted as "give me the process needed for the batch" instead of "go an actually process the batch". So I think `process_batch` would be better.

vedanshujain replied:
👍  renamed in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-1e070cc170d3b65cd6ba61b60cbd0697c8431fda674b4720926c8fbc1afecce7R297

Konamiman commented on plugins/woocommerce/src/Internal/Utilities/BatchProcessor.php:
```diff
@@ -0,0 +1,66 @@
+	/**
+	 * Executed by updatecontroller when update is complete. Child classes may want to do custom action here, such as showing notices if they want to.
+	 */
+	public function mark_process_complete() {
```
I still think that this can go away. A processor can know, at the end of `process_for_batch`, if there's still more data to be processed or not, and act consequently (especially if we pass state data as I suggested for `get_batch_data`). Additionally, removing this would allow to convert the abstract class into an interface.

vedanshujain replied:
Hmm, I have refactored a bit to remove this method and changed the class to interface as suggested in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-06503a1236948e2273ef4e52827e4509e27e1c2d699c0a7151f1d2321919bdebR13. Looks nicer now, thanks!

Konamiman commented on plugins/woocommerce/src/Internal/Utilities/BatchProcessingController.php:
```diff
@@ -0,0 +1,287 @@
+	 *
+	 * @return mixed Update status.
+	 */
+	protected function get_process_details( BatchProcessor $batch_processor ) {
```
Do this method and the companion `update_progress_status` really need to be protected? I can't think of any use case for "messing" with the batch enqueueing/scheduling process in a derived class, and if what you want is to provide visibility for the sake of logging/data visualization, I think that a couple of hooks would do a better job here.

vedanshujain replied:
Agreed, modified to make then `private` in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-77d9531091cce5f175d5bb8fe3a1560d9dda936d370c81fabb90ca0f46633e52R170

Konamiman commented on plugins/woocommerce/src/Internal/Utilities/BatchProcessingController.php:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	/**
+	 * Helper method to get list of all pending processes.
```
Documentation improvement suggestion: what's this method going to return exactly? What's the shape of each of the items in the returned array?

vedanshujain replied:
Added some documentation in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-77d9531091cce5f175d5bb8fe3a1560d9dda936d370c81fabb90ca0f46633e52R231

Konamiman commented on plugins/woocommerce/src/Internal/Utilities/BatchProcessingController.php:
```diff
@@ -0,0 +1,287 @@
+	 *
+	 * @param string $processor Name of processor class which will be marked compete.
+	 */
+	protected function mark_pending_process_complete( string $processor ) {
```
Again, could this perhaps use a hook instead of being a protected method?

vedanshujain replied:
I have marked this method as private for now. As you said in an earlier conversation, cleanup action can be done after checking if there is no batch remaining.

Konamiman commented:
Thanks for the changes, and what you say about my "Enqueueing vs registering" and "Additional hooks" comments makes sense. I still miss the "Robustness" part, though; what's going to happen if a batch times out or throws an exception? Won't that interrupt the entire process?

Konamiman commented:
One more thing: maybe these new classes should go into a separate directory/namespace, e.g. `Internal\BatchProcessing`. `Utilities` is more suited for classes for small and often static utility methods to be used freely across the codebase.

vedanshujain committed:
PR feedback changes.

vedanshujain committed:
Removed unused method.

vedanshujain committed:
Address PR feedback.

vedanshujain committed:
Applied coding standards.

vedanshujain commented:
> I still miss the "Robustness" part, though; what's going to happen if a batch times out or throws an exception?

So to make things robust, we currently run two processes, one is the normal batch process `SINGLE_BATCH_PROCESS_ACTION` which schedules itself after processing the batch. However, we run another process `CONTROLLER_CRON_NAME` which periodically checks if whether there is a pending process that is not scheduled, (and schedules if not.

vedanshujain committed:
Ignore todo statmement phpcs error.

vedanshujain commented:
hey @Konamiman this is ready for another review (subjected to tests passing)

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,300 @@
+
+namespace Automattic\WooCommerce\Internal\BatchProcessing;
+
+use Dompdf\Exception;
```
This happens to me all the time too, the IDE tools are tricky 😄

vedanshujain replied:
Nice catch, removed in https://github.com/woocommerce/woocommerce/pull/33233/commits/311f014dc39a5cfe7be85e3eef6c332cddae4618

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,300 @@
+	 */
+	private function get_args_for_init_cron() {
+		return array();
+	}
```
I think this method can go away and just use the empty array directly, right?

vedanshujain replied:
Agreed, removed in https://github.com/woocommerce/woocommerce/pull/33233/commits/311f014dc39a5cfe7be85e3eef6c332cddae4618

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,300 @@
+		/**
+		 * Filters the instance of processor for current class name.
+		 *
+		 * @since 6.7.0.
```
I think this is going to be more like 6.8.0 at this point.

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,300 @@
+	 *
+	 * @param string $processor_class_name Fully qualified class name of the processor, must be child class of `BatchProcessor`.
+	 */
+	public function enqueue_processor( string $processor_class_name ) {
```
I have a question about this. Imagine this scenario:

1. `enqueue_processor` is invoked, this schedules `CONTROLLER_CRON_NAME` to be run immediately
2. `schedule_processes` runs, this schedules `CONTROLLER_CRON_NAME` again but for one hour later
3. Before the hour passes, `enqueue_processor` is invoked, scheduling `CONTROLLER_CRON_NAME`again

What happens then? The last scheduled action overrides the first one, or both keep scheduled and run at their proper times? If so, shouldn't we be checking if the action is already scheduled before scheduling a new one, to prevent a neverending stream of "action lanes" (one per execution of `enqueue_processor`) that would keep rescheduling themselves over and over?

vedanshujain replied:
You are right, seems like I had missed this. I have added a `unique` param in https://github.com/woocommerce/woocommerce/pull/33233/commits/4be2dd5e3d88ed3883deb5aa41ac2e1a2ec98982 when we first start the process from `enqueue_processor` (but not on the re-schedule ones since they are still running and will be excluded). Note that this is a fairly new API that has not been released yet, although it's merged, so if you want to test you would have checkout the latest trunk in Action Scheduler.

Konamiman commented:
> So to make things robust, we currently run two processes, one is the normal batch process SINGLE_BATCH_PROCESS_ACTION which schedules itself after processing the batch. However, we run another process CONTROLLER_CRON_NAME which periodically checks if whether there is a pending process that is not scheduled, (and schedules if not.

So if I understood correctly:

1. If a batch processing throws an exception, that's going to be caught and logged, then the next processing will be scheduled as usual.
2. If a batch processing causes the action to timeout, the `CONTROLLER_CRON_NAME` will detect that in at most one hour, and will reschedule the action.

Even then, I think it would be convenient to detect timeouts (I guess AS has some dedicated API for that?) and maybe trigger a dedicated action passing the processor class name and the offending batch, so that controllers can throttle back the batch size.

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,300 @@
+	 *
+	 * @param BatchProcessorInterface $batch_processor Batch processor instance.
+	 *
+	 * @return bool True if batch was processed, false if not.
```
This doesn't seem to be used, can we remove it and just return nothing?

vedanshujain replied:
Removed in https://github.com/woocommerce/woocommerce/pull/33233/commits/311f014dc39a5cfe7be85e3eef6c332cddae4618

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,300 @@
+			$processor = new $processor_class_name();
+		}
+		if ( ! is_a( $processor, BatchProcessorInterface::class ) ) {
+			throw new \Exception( 'Unable to initialize batch processor instance.' );
```
Let's add the class name to the exception message to make troubleshooting easier?

vedanshujain replied:
👍  Done in https://github.com/woocommerce/woocommerce/pull/33233/commits/311f014dc39a5cfe7be85e3eef6c332cddae4618

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,300 @@
+	 */
+	protected function log_error( \Exception $error ) : void {
+		// phpcs:ignore
+		// TODO: Implement error logging.
```
Should we implement this? The batch processor itself and the batch at hand should probably be passed to the method too, so we can log something like _Batch processing failed, processor: (class name), batch: (batch items), error: (exception message)_ - batch items to be included only if they are all strings or numbers.

vedanshujain replied:
Added implementation in https://github.com/woocommerce/woocommerce/pull/33233/commits/a74574a1da2529c54594e66509b30123b989ad24

vedanshujain committed:
Make scheduling initial actions unique.

vedanshujain committed:
Address PR feedback.

vedanshujain committed:
Add logging.

vedanshujain committed:
Better error handling, also throw error to AS when a process failed.

Konamiman commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -0,0 +1,323 @@
+	 * @param array                   $batch Batch that finished processing.
+	 */
+	protected function log_error( \Exception $error, BatchProcessorInterface $batch_processor, array $batch ) : void {
+		$batch_detail_string = '';
```
Suggestion for composing this string:

1. Check if all the batch items are numbers, and if so, use [ArrayUtil::to_ranges_string](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Utilities/ArrayUtil.php#L104)
2. If not, check if first and last items are numbers or strings, and include them.
3. If not, revert to `print_r` or maybe even do not include nothing.

We can add some documentation recommending using numbers whenever possible, or strings if not, for the returned batches.

vedanshujain replied:
I wonder if going through all data in batches and type checking will be a bit of a performance issue? I have added a filter on the final log message in https://github.com/woocommerce/woocommerce/pull/33233/commits/efb29e540a8e37f2507172d57d4f18582f55cf0b, maybe that's enough for this PR. wdyt?

Konamiman replied:
Keep in mind that this code is going to run only in case of error, so it shouldn't run often and it it runs it means that something is broken and thus performance won't really matter. But sure, a filter will do for now.

vedanshujain committed:
Remove callee expectation from total pending count.

vedanshujain committed:
Add a filter to modify error message in logs.

Konamiman committed:
Update code comments, rename methods/parameters, add return types.

Konamiman commented:
I think this is ready to go as soon as the errors in the build are fixed (if these are really bugs in this PR)

vedanshujain committed:
Make the method declaration compatible with parent interface.
